### PR TITLE
V-BLAST

### LINKED ISSUE #10: Create encoder block
Create C++ encoder block.

Block: general
Input (1 port): complex vectors
Output (M ports): complex vectors, M = 2,4
Args: M, vector_length

### LINKED ISSUE #11: Create decoder block
Create C++ decoder block. The implemented decoding techniques are the linear approaches **Zero Forcing** (ZF) and **Minimum Mean Square Error** (MMSE).

* Block: general
* Input (N ports): complex vectors
* Output (1 port): complex vectors
* Args: N, vector length, decoding mode (ZF, MMSE)
* CSI: Transported over stream tags.

### LINKED ISSUE #12: QA Tests for V-BLAST
Create a qa test in python with a bunch tests, providing confidence that quality requirements will be fulfilled. Additionally a loopback test with encoder and decoder is provided (with known CSI).

### PR DESCRIPTION
**Pull request by MLsmd**
- Implement V-BLAST encoder block. This closes #10 .
- Implement V-BLAST decoder block. This closes #11 .
- Add python qa tests for the C++ blocks and loop back test. This closes #12 .
- Create .xml files for GRC.
- Documentation for Doxygen.

### CODE REVIEW

**Comment by swap-nil7** [comment]
In my opinion, the `grc` XML file should be replaced by a YAML file as there is no XML support in the `next`. XML is okay for `version 3.7` though.

**Comment by MLsmd** [comment]
@swap-nil7 thanks for the info! But I am currently developing for GNU Radio version 3.7. After the project is finished and merged into master, the XML files can be converted to YAML files with the rest of the GNU Radio modules from 3.7 with some kind of automated scripts.

File: gr-digital/include/gnuradio/digital/vblast_decoder_cc.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  namespace digital {
+
+/*! \brief VBLAST decoder.
+ * Decodes incoming MxM MIMO data by equalizing the received signals in order to extract the
+ * M different transmission signals. For the equalization, you can choose between a
+ * zero forcing (key='ZF') and a minimum mean squared error (key='MMSE') scheme. Both schemes require
+ * channel state information and MMSE additionally needs SNR information.
+ *
+ * The CSI and SNR information is transported via stream tags with key='csi' or 'snr', respectively.
+ * Initially the CSI is set to 1.0 + 0j for both branches and are updated
```
Comment (fwunsch) [inline gr-digital/include/gnuradio/digital/vblast_decoder_cc.h]: "both" branches?

File: gr-digital/include/gnuradio/digital/vblast_decoder_cc.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ * with each incoming CSI. The SNR is initially set to 1.0e6; in this case is the MMSE equalizer equal
+ * to the ZF equalizer. The CSI and SNR tags are processed separately and they can therefore arrive
+ * at different positions of the stream and occur in different frequencies.
```
Comment (fwunsch) [inline gr-digital/include/gnuradio/digital/vblast_decoder_cc.h]: "in" -> "at"

File: gr-digital/include/gnuradio/digital/vblast_encoder_cc.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+#include <gnuradio/digital/api.h>
+#include <gnuradio/sync_decimator.h>
+
+namespace gr {
+  namespace digital {
+
+/*! \brief VBLAST encoder.
+ * Encodes a serial input data stream into a vertical transmission vector
+ * which is transmitted at once over multiple antennas.
+ * The functionality equals a multiplexing of the data.
```
Comment (fwunsch) [inline gr-digital/include/gnuradio/digital/vblast_encoder_cc.h:36]: Isn't this what the GR block "stream to streams" does? I think we need to discuss with @marcusmueller if we need a second block doing the exact same thing just to have a nicely named block.

**Review by fwunsch — CHANGES_REQUESTED** [review]
Nice work! Only details from my side. 

One thing needs to be discussed, though: You implement the VBLAST encoder which basically could be replaced by the stream to streams block that is already in GR. While on the one hand it makes the MIMO functionality independent from possible future changes to the GR block and gives the opportunity to have a proper and descriptive name for the encoder, it basically duplicates functionality. Pinging @marcusmueller with this.

File: gr-digital/include/gnuradio/digital/vblast_encoder_cc.h (hunk 3 above)
Comment (MLsmd) [reply gr-digital/include/gnuradio/digital/vblast_encoder_cc.h:36]: I can see you point. Both options have their eligibility.

**Comment by MLsmd** [comment]
@marcusmueller Ready for merging.

**Review by fwunsch — APPROVED** [review]

**Pull request merged**